src/System.Memory/src/System/Buffers/ReadOnlySequence_helpers.cs

```diff
@@ -14,105 +14,95 @@ namespace System.Buffers
+                if (positionSegment != _sequenceEnd.GetObject())
                 {
-                    ReadOnlySequenceSegment<T> nextSegment = startSegment.Next;
 
-                    if (nextSegment == null && startSegment != endObject)
-                        ThrowHelper.ThrowInvalidOperationException_EndPositionNotReached();
+                    ReadOnlySequenceSegment<T> nextSegment = positionSegment.Next;
+                    Debug.Assert(nextSegment != null);
 
                     next = new SequencePosition(nextSegment, 0);
-                    length = data.Length - startIndex;
+                    endIndex = memory.Length;
                 }
             }
-            else if (type == SequenceType.Array)
+            else
             {
-                Debug.Assert(startObject is T[]);
+                // Bounds check
+                Debug.Assert(positionObject == _sequenceEnd.GetObject());
+                Debug.Assert(GetIndex(_sequenceStart) <= GetIndex(position) && GetIndex(position) <= GetIndex(_sequenceEnd));
 
-                data = new ReadOnlyMemory<T>(Unsafe.As<T[]>(startObject));
-            }
-            else if (type == SequenceType.MemoryManager)
-            {
-                Debug.Assert(startObject is MemoryManager<T>);
+                if (type == SequenceType.Array)
+                {
+                    Debug.Assert(positionObject is T[]);
 
-                data = (Unsafe.As<MemoryManager<T>>(startObject)).Memory;
-            }
-            else if (typeof(T) == typeof(char) && type == SequenceType.String)
-            {
-                Debug.Assert(startObject is string);
+                    memory = new ReadOnlyMemory<T>(Unsafe.As<T[]>(positionObject));
+                }
+                else if (typeof(T) == typeof(char) && type == SequenceType.String)
+                {
+                    Debug.Assert(positionObject is string);
 
-                data = (ReadOnlyMemory<T>)(object)(Unsafe.As<string>(startObject)).AsMemory();
-            }
-            else
-            {
-                data = default;
-                return false;
+                    memory = (ReadOnlyMemory<T>)(object)(Unsafe.As<string>(positionObject).AsMemory());
+                }
+                else // if (type == SequenceType.MemoryManager)
+                {
+                    Debug.Assert(positionObject is MemoryManager<T>);
+
+                    memory = Unsafe.As<MemoryManager<T>>(positionObject).Memory;
+                }
             }
 
-            data = data.Slice(startIndex, length);
+            int positionIndex = GetIndex(position);
+            memory = memory.Slice(positionIndex, endIndex - positionIndex);
             return true;
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        internal ReadOnlyMemory<T> GetFirstBuffer(in SequencePosition start, in SequencePosition end)
+        private ReadOnlyMemory<T> GetFirstBuffer()
```

Comment:
Try something like this:
```C#
[MethodImpl(MethodImplOptions.AggressiveInlining)]
private ReadOnlyMemory<T> GetFirstBuffer()
{
    object startObject = _sequenceStart.GetObject();

    if (startObject == null)
    {
        return default;
    }

    SequenceType type = GetSequenceType();
    object endObject = _sequenceEnd.GetObject();
    int startIndex = GetIndex(_sequenceStart);
    int endIndex = GetIndex(_sequenceEnd);

    if (type == SequenceType.MultiSegment)
    {
        Debug.Assert(startObject is ReadOnlySequenceSegment<T>);

        ReadOnlySequenceSegment<T> startSegment = Unsafe.As<ReadOnlySequenceSegment<T>>(startObject);

        if (startObject != endObject)
        {
            return startSegment.Memory.Slice(startIndex);
        }
        return startSegment.Memory.Slice(startIndex, endIndex - startIndex);
    }

    if (type == SequenceType.Array)
    {
        if (startObject != endObject)
            ThrowHelper.ThrowInvalidOperationException_EndPositionNotReached();

        Debug.Assert(startObject is T[]);

        return new ReadOnlyMemory<T>(Unsafe.As<T[]>(startObject), startIndex, endIndex - startIndex);
    }

    if (typeof(T) == typeof(char) && type == SequenceType.String)
    {
        if (startObject != endObject)
            ThrowHelper.ThrowInvalidOperationException_EndPositionNotReached();

        Debug.Assert(startObject is string);

        return (ReadOnlyMemory<T>)(object)(Unsafe.As<string>(startObject)).AsMemory(startIndex, endIndex - startIndex);

    }

    Debug.Assert(type == SequenceType.MemoryManager);

    if (startObject != endObject)
        ThrowHelper.ThrowInvalidOperationException_EndPositionNotReached();

    Debug.Assert(type == SequenceType.MemoryManager);
    Debug.Assert(startObject is MemoryManager<T>);

    return Unsafe.As<MemoryManager<T>>(startObject).Memory.Slice(startIndex, endIndex - startIndex);
}
```

```C#
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        private SequenceType GetSequenceType()
        {
            return (SequenceType)(-(2 * (_sequenceStart.GetInteger() >> 31) + (_sequenceEnd.GetInteger() >> 31)));
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        private SequenceType GetSequenceType(in SequencePosition position)
        {
            return (SequenceType)(-(2 * (position.GetInteger() >> 31) + (_sequenceEnd.GetInteger() >> 31)));
        }
```

![image](https://user-images.githubusercontent.com/6527137/38415517-9ca51f30-3946-11e8-8436-81eea19c20b5.png)